### PR TITLE
Add ring buffer

### PR DESCRIPTION
**Pull request by dandelion-james**
Ring buffer.
Fixed size container, always full (initially with default constructed elements).
Methods: begin, end, and push. Pushed elements overwrite the oldest elements.
Intended to be used with ProfilerGraph, to replace deque m_log.
Implements a forward iterator.
Could be enhanced to support random access iteration and const_iterator.

### CODE REVIEW

**Comment by gregorycu** [comment]
I wouldn't have made the buffer size a template parameter.

Also, in the ring buffer .size() is always equal to .max_size(). I may have not implemented it like that.

A potentially different design would be to require that the ring buffer have a reserve() member that sets the minimum size. And push() could grow the ring (if required) and pop() would remove elements (but not reduce the size, if it could).

So basically acts like the std::vector, you can resize(), and from then on it will retain the size unless you add too many and don't remove enough.

Good code though.

**Comment by est31** [comment]
I wouldn't have made the buffer size a parameter either. I'd have passed the size in the constructor, that way it can be chosen at run-time.

Also, I don't think we need a separate iterator type. A simple `T get(n)` method would be enough, I think.

Otherwise, looks okay.

File: src/util/container.h [hunk 1]
```diff
@@ -303,5 +304,76 @@ class LRUCache
...
+		return !(*this == other);
+	}
+private:
+	size_t m_offset;
+	T m_data[N];
+friend class RingBufferIter<T, N>;
+};
+
+template <typename T, size_t N>
+class RingBufferIter { // TODO: random access iteration (if needed)
```
Comment (ShadowNinja) [inline src/util/container.h]: Inherit from `std::iterator`?

File: src/util/container.h [hunk 2]
```diff
@@ -303,5 +304,76 @@ class LRUCache
...
+				return false;
+		return true;
+	}
+	bool operator!=(const RingBuffer& other) const {
+		return !(*this == other);
+	}
+private:
+	size_t m_offset;
+	T m_data[N];
+friend class RingBufferIter<T, N>;
```
Comment (ShadowNinja) [inline src/util/container.h]: Should be indented and separated by an empty line.

**Comment by dandelion-james** [comment]
I agree that the iterator is unnecessary. I'll replace it with access via at (in imitation of std).

In defense of compile-time size:
I chose the less flexible solution in the hopes that the compiler can make good use of the information. The flexibility isn't needed in this case. The container can be modified as the need arises.

File: src/util/container.h [hunk 3]
```diff
@@ -303,5 +304,54 @@ class LRUCache
...
+	
+	T& at(size_t n)
+	{
+		if (n >= N)
+			throw std::out_of_range("RingBuffer::at");
+		if (n + m_offset < N)
+			return m_data[n + m_offset];
+		else
+			return m_data[n + m_offset - N];
+	}
```
Comment (ShadowNinja) [inline src/util/container.h]: It seems that pushing a value causes the indexes for at() to shift over by one, which seems like an odd way to do things.

Comment (est31) [inline src/util/container.h]: Agreed, its different from `vector` behaviour, where the value from `at(0)` always points to the first element you have `push()`-ed.

Comment (gregorycu) [inline src/util/container.h]: I don't know. I think that at() should always point to the "oldest" element (this is the element that would be overridden with a push()). The ring buffer mimics a fixed size queue where you would expect the element address returned by at(0) to change. With vector, at(0) points to the first element in the sequence. So does this one. This implementation is not odd at all.

Comment (est31) [inline src/util/container.h]: No, the behaviour here is that at(0) points to the element last pushed.

Comment (gregorycu) [inline src/util/container.h]: push has:
m_data[m_offset] = t;

Which means that at(0) points to the oldest position (the next one to be erased)... Right?

Comment (est31) [inline src/util/container.h]: push has:
m_data[m_offset] = t;

at(n) has:
return m_data[n + m_offset];

**Comment by est31** [comment]
:+1:

**Comment by est31** [comment]
Can you add unit tests? Edit `src/unittest/test_utilities.cpp` for it.

**Comment by paramat** [comment]
The author has deleted their repo, this would perhaps need to be taken over by someone else.

**Comment by paramat** [comment]
Will close in 1 month if no-one wnats to take over on this.

**Comment by Fixer-007** [comment]
Minetest Never Changes 。。。

**Comment by paramat** [comment]
Fixer never changes ...

**Comment by Fixer-007** [comment]
:coffin:

**Comment by nerzhul** [comment]
i don't see interest to add unused code

**Comment by Zeno-** [comment]
This should be closed. If it's intended "to be used with ProfilerGraph, to replace deque m_log" then that should be done and this change part of that.

**Comment by Fixer-007** [comment]
Something is really wrong with minetest development or with this guy or both...

http://irc.minetest.net/minetest-dev/2015-11-09#i_4446114

>00:31 	dandelion1 	Hey, I'm new. Would someone recommend a little maintenance issue to get me started?
00:32 	est31 	seems 4gb aren't enough to have stk + firefox with a moderate number open + hexchat
00:32 		dandelion1, a starter issue?
00:32 	dandelion1 	Yeah
00:33 	est31 	dandelion1, which part of minetest are you interested in helping in? minetest_game, minetest engine? client? server?
00:34 	dandelion1 	Engine. Eventually I want to do sqlite or threading
00:37 		Those two things I'm new to. I've done a lot with fractals/cellular automata/dungeon generation
00:47 	est31 	dandelion1, do you know how to use c++ templates?
00:47 	dandelion1 	Yes
00:47 	est31 	nice.
00:48 		can you write a ring buffer type and place it in src/util/container.h ?
00:48 	dandelion1 	Okay
00:48 	est31 	& then use it to speed up on gregorycu's speedup of the profiler code
00:50 		the class is called ProfilerGraph and resides in src/game.cpp
00:50 		idea is to replace m_log with a ring buffer

ONE YEAR passed... Happy End...

http://irc.minetest.net/minetest-dev/2016-10-26#i_4731106

>04:38 	ShadowBot 	https://github.com/minetest/minetest/issues/3354 -- Add ring buffer by dandelion-james
04:38 	Zeno` 	it doesn't do anything
04:39 	sofar 	was there even unit tests in that?
04:39 	Zeno` 	no
04:39 		my comment on closing is "If it's intended "to be used with ProfilerGraph, to replace deque m_log" then that should be done and this change part of that."
04:39 	sofar 	well that's surely a sign that it doesn't do anything :)
04:39 	Zeno` 	as that was the author's justification for the PR
04:43 		hehe
04:44 		well, yeah. Not much point in adding dead code before it even lives

**Comment by est31** [comment]
Thanks @Fixer-007 for digging this up! I knew there was an important reason for this PR but forgot why and didn't want to do the digging. We should try to get it merged and then use it like I described one year ago.

And yeah, its really sad that this PR sat for so long. Maybe I should have pushed it a little more...

**Comment by sofar** [comment]
I don't think this is necessarily a failure of reviewers. The original submitter got lots of reviews in the first week on this PR, but seems to have vanished entirely. His github page shows he basicallly hasn't touched github since making the last comment on this page 1 day after he submitted the PR.

I doubt that the original submitter left because of minetest reviews in this case.

Yes, it sucks if good code is lost because of it, but we're a community, and people do come and go as they please.

**Comment by paramat** [comment]
Yeah the author didn't reply to a review or the question about unit tests, could have bumped this or chatted on IRC.
